src/geometry/euclidean.lean

```diff
@@ -33,3 +35,634 @@ normed_add_torsor V P
+                          (abs_le.mp (abs_inner_div_norm_mul_norm_le_one x y)).2
+                          (by norm_num)
+                          (by norm_num)
+                          h },
+  { intro h,
+    rw h }
+end
+
+/-- If the angle between two vectors is π, the angles between those
+vectors and a third vector add to π. -/
+lemma angle_of_vectors_add_angle_of_vectors_eq_pi_of_angle_of_vectors_eq_pi {x y : V} (z : V)
+    (h : angle_of_vectors x y = real.pi) :
+  angle_of_vectors x z + angle_of_vectors y z = real.pi :=
+begin
+  rw angle_of_vectors_eq_pi_iff at h,
+  rcases h with ⟨hx, ⟨r, ⟨hr, hxy⟩⟩⟩,
+  rw [hxy, angle_of_vectors_of_neg_smul_left x z hr, angle_of_vectors_of_neg_left,
+      add_sub_cancel'_right]
+end
+
+/-- Two vectors have inner product 0 if and only if the angle between
+them is π/2. -/
+lemma inner_eq_zero_iff_angle_eq_pi_div_two (x y : V) :
+  inner x y = 0 ↔ angle_of_vectors x y = real.pi / 2 :=
+begin
+  split,
+  { intro h,
+    unfold angle_of_vectors,
+    rw [h, zero_div, real.arccos_zero] },
+  { intro h,
+    unfold angle_of_vectors at h,
+    rw ←real.arccos_zero at h,
+    have h2 : inner x y / (∥x∥ * ∥y∥) = 0 :=
+      real.arccos_inj (abs_le.mp (abs_inner_div_norm_mul_norm_le_one x y)).1
+                      (abs_le.mp (abs_inner_div_norm_mul_norm_le_one x y)).2
+                      (by norm_num)
+                      (by norm_num)
+                      h,
+    by_cases h : (∥x∥ * ∥y∥) = 0,
+    { cases eq_zero_or_eq_zero_of_mul_eq_zero h with hx hy,
+      { rw norm_eq_zero at hx,
+        rw [hx, inner_zero_left] },
+      { rw norm_eq_zero at hy,
+        rw [hy, inner_zero_right] } },
+    { rwa div_eq_zero_iff h at h2 } },
+end
+
+/-- Pythagorean theorem, if-and-only-if vector inner product form. -/
+lemma norm_add_square_eq_norm_square_add_norm_square_iff_inner_eq_zero (x y : V) :
+  ∥x + y∥ * ∥x + y∥ = ∥x∥ * ∥x∥ + ∥y∥ * ∥y∥ ↔ inner x y = 0 :=
+begin
+  rw [norm_add_mul_self, add_right_cancel_iff, add_right_eq_self, mul_eq_zero],
+  norm_num
+end
+
+/-- Pythagorean theorem, subtracting vectors, if-and-only-if vector
+inner product form. -/
+lemma norm_sub_square_eq_norm_square_add_norm_square_iff_inner_eq_zero (x y : V) :
+  ∥x - y∥ * ∥x - y∥ = ∥x∥ * ∥x∥ + ∥y∥ * ∥y∥ ↔ inner x y = 0 :=
+begin
+  rw [norm_sub_mul_self, add_right_cancel_iff, sub_eq_add_neg, add_right_eq_self, neg_eq_zero,
+      mul_eq_zero],
+  norm_num
+end
+
+/-- Pythagorean theorem, if-and-only-if vector angle form. -/
+lemma norm_add_square_eq_norm_square_add_norm_square_iff_angle_eq_pi_div_two (x y : V) :
+  ∥x + y∥ * ∥x + y∥ = ∥x∥ * ∥x∥ + ∥y∥ * ∥y∥ ↔ angle_of_vectors x y = real.pi / 2 :=
+begin
+  rw norm_add_square_eq_norm_square_add_norm_square_iff_inner_eq_zero,
+  exact inner_eq_zero_iff_angle_eq_pi_div_two x y
+end
+
+/-- Pythagorean theorem, subtracting vectors, if-and-only-if vector angle form. -/
+lemma norm_sub_square_eq_norm_square_add_norm_square_iff_angle_eq_pi_div_two (x y : V) :
```

Comment:
These statements are useful! But if I want to rewrite the equality, I need to write
```
(norm_sub_square_eq_norm_square_add_norm_square_iff_angle_eq_pi_div_two x y).mpr h
```
which is quite a mouthful.
Could you please add
`norm_sub_square_eq_norm_square_add_norm_square` (and a version with a `'`) that take the assumption on the angle/inner as hypothesis?

Reply:
Added four such variants of Pythagoras (for vectors, so there are now eight variants for vectors and one for points; once triangles are defined there will need to be at least one for triangles as well).